src/model/position.js

```diff
@@ -1057,3 +1056,22 @@ export default class Position {
  *
  * @typedef {String} module:engine/model/position~PositionStickiness
  */
+
+// This helper method concatenate two arrays more efficiently than Array.concat(). See ckeditor/ckeditor5#6528.
+//
+// The problem with Array.concat() is that it is an overloaded method that can concatenate various arguments,
+// like mixed data types with arrays (e.g. [ 0 ].concat( 1, 2, [3, 4])) thus it probably check each argument's types and more.
+// In Position's paths concatenation case there always be two arrays to merge so such general method is not needed.
+function concatenatePaths( rootPath, path ) {
+	const newPath = [];
```

Comment:
Did you check whether creating new array with the desired length (\`new Array(len)\`) isn't better?

Creating an empty array and pushing items to it seems extremely counterintuitive. I understand that right now we can see that it's faster, but this is this type of optimization that may suddenly become slower. 

Also, on which browsers did you test it? Due to this being counterintuitive, I'd like to be 100% sure that it doesn't deoptimize this on one of the browsers. It's probably one of the places where having a manual test comparing two scenarios might be helpful in verifying this in the future too. Because it's not said that the situation won't change.

Reply:
> Did you check whether creating new array with the desired length (`new Array(len)`) isn't better?

I'll check. The same thing was pointed out by @mlewand . I think that in this case, the change might be negligible as both arrays have length of max few values. This is potentially helpful when concatenating large arrays.

The case that this might get slower will probably be a newer a case for `Array.concat()` as this method do too many things AFAICS (must check each parameter from `args` if it is an array, etc).

Responding also to `spread` operator. It _is_ slower. I'll post the results below. Also, I'll add more comments in the code so one would remember why `for` and not `[ ...root, ...path ]`.

Reply:
Anyway, I'll check this with a manual test and post the results.

Reply:
So the results are

### 1\. Proposed solution

![](https://user-images.githubusercontent.com/247363/78563910-f8b0ae00-781b-11ea-82e1-f2b587862220.png)

![](https://user-images.githubusercontent.com/247363/78563922-fc443500-781b-11ea-9576-f9cd7c928049.png)

### 2\. Spread operator:

![](https://user-images.githubusercontent.com/247363/78563962-0c5c1480-781c-11ea-8935-90cf3972c03d.png)

![](https://user-images.githubusercontent.com/247363/78563997-1a119a00-781c-11ea-8d1e-515c99a4f8a7.png)

### 3\. Ultra optimized array creation

![](https://user-images.githubusercontent.com/247363/78564093-3d3c4980-781c-11ea-9acb-bd11ca66cc66.png)

![](https://user-images.githubusercontent.com/247363/78564198-5d6c0880-781c-11ea-852e-9972d98b6184.png)

Reply:
On Firefox (using medium sample because FF has short buffer and it way to slow there):

### 1\. Proposed

![](https://user-images.githubusercontent.com/247363/78565100-b38d7b80-781d-11ea-8c1e-4a25eb378585.png)

### 2\. Spread operator

![](https://user-images.githubusercontent.com/247363/78565134-bdaf7a00-781d-11ea-9333-d0c5068b9e6f.png)

It looks like there is no difference there but number of samples is different so it looks like the performance buffer was overflown.

Reply:
On Firefox (using medium sample because FF has short buffer and it way to slow there):

### 1\. Proposed

![](https://user-images.githubusercontent.com/247363/78565100-b38d7b80-781d-11ea-8c1e-4a25eb378585.png)

### 2\. Spread operator

![](https://user-images.githubusercontent.com/247363/78565134-bdaf7a00-781d-11ea-9333-d0c5068b9e6f.png)

It looks like there is no difference there but number of samples is different so it looks like the performance buffer was overflown.

Reply:
So 1 & 3 are basically the same so we can choose either.